cns/deviceplugin/pluginmanager.go

```diff
@@ -0,0 +1,147 @@
+package deviceplugin
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/Azure/azure-container-networking/crd/multitenancy/api/v1alpha1"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"k8s.io/kubelet/pkg/apis/deviceplugin/v1beta1"
+)
+
+var ErrUnknownDeviceType = errors.New("unknown device type")
+
+const (
+	defaultDevicePluginDirectory = "/var/lib/kubelet/device-plugins"
+	defaultDeviceCheckInterval   = 5 * time.Second
+)
+
+type pluginManagerOptions struct {
+	devicePluginDirectory string
+	kubeletSocket         string
+	deviceCheckInterval   time.Duration
+}
+
+type pluginManagerOption func(*pluginManagerOptions)
+
+func PluginManagerSocketPrefix(prefix string) func(*pluginManagerOptions) {
+	return func(opts *pluginManagerOptions) {
+		opts.devicePluginDirectory = prefix
+	}
+}
+
+func PluginManagerKubeletSocket(socket string) func(*pluginManagerOptions) {
+	return func(opts *pluginManagerOptions) {
+		opts.kubeletSocket = socket
+	}
+}
+
+func PluginDeviceCheckInterval(i time.Duration) func(*pluginManagerOptions) {
+	return func(opts *pluginManagerOptions) {
+		opts.deviceCheckInterval = i
+	}
+}
+
+// PluginManager runs device plugins for vnet nics and ib nics
+type PluginManager struct {
```

Comment:
I weighed the pros and cons of having a more generic "plugin manager" that manages any kind of plugin (and even started coding it that way), but settled on this coupling because it keeps the management of the socket file all in one place. Not strongly tied to this approach but it felt simpler with no strong benefit to keeping it generic (probably are never going to add another device type to this)

Reply:
That said I'm ok with making this generic but there will be some refactoring involved, specifically around letting the plugins clean up any bad state that was left behind as part of `Run`

Reply:
this is still a significant refactoring of plugin-manager, for the sake of making this generic before we have a second plugin/scenario to validate that design. 

I don't think this needs to be blocking the PR to add device-plugin functionality at this point - we can keep this simple, and consider the refactoring if/when we have a concrete reason to extend this.